pkg/kubelet/apis/kubeletconfig/v1beta1/types.go

```diff
@@ -40,303 +40,473 @@ const (
 	//   "promiscuous-bridge": make the container bridge promiscuous.
 	//   "hairpin-veth":       set the hairpin flag on container veth interfaces.
 	//   "none":               do nothing.
 	// Generally, one must set --hairpin-mode=hairpin-veth to achieve hairpin NAT,
 	// because promiscous-bridge assumes the existence of a container bridge named cbr0.
-	HairpinMode string `json:"hairpinMode"`
+	// Default: "promiscuous-bridge"
+	// +optional
+	HairpinMode string `json:"hairpinMode,omitempty"`
 	// maxPods is the number of pods that can run on this Kubelet.
-	MaxPods int32 `json:"maxPods"`
+	// Default: 110
+	// +optional
+	MaxPods int32 `json:"maxPods,omitempty"`
 	// The CIDR to use for pod IP addresses, only used in standalone mode.
 	// In cluster mode, this is obtained from the master.
-	PodCIDR string `json:"podCIDR"`
+	// Default: ""
+	// +optional
+	PodCIDR string `json:"podCIDR,omitempty"`
 	// ResolverConfig is the resolver configuration file used as the basis
 	// for the container DNS resolution configuration.
-	ResolverConfig string `json:"resolvConf"`
-	// cpuCFSQuota is Enable CPU CFS quota enforcement for containers that
-	// specify CPU limits
-	CPUCFSQuota *bool `json:"cpuCFSQuota"`
+	// Default: "/etc/resolv.conf"
+	// +optional
+	ResolverConfig string `json:"resolvConf,omitempty"`
+	// cpuCFSQuota enables CPU CFS quota enforcement for containers that
+	// specify CPU limits.
+	// Default: true
+	// +optional
+	CPUCFSQuota *bool `json:"cpuCFSQuota,omitempty"`
 	// maxOpenFiles is Number of files that can be opened by Kubelet process.
-	MaxOpenFiles int64 `json:"maxOpenFiles"`
+	// Default: 1000000
+	// +optional
+	MaxOpenFiles int64 `json:"maxOpenFiles,omitempty"`
 	// contentType is contentType of requests sent to apiserver.
-	ContentType string `json:"contentType"`
+	// Default: "application/vnd.kubernetes.protobuf"
+	// +optional
+	ContentType string `json:"contentType,omitempty"`
 	// kubeAPIQPS is the QPS to use while talking with kubernetes apiserver
-	KubeAPIQPS *int32 `json:"kubeAPIQPS"`
-	// kubeAPIBurst is the burst to allow while talking with kubernetes
-	// apiserver
-	KubeAPIBurst int32 `json:"kubeAPIBurst"`
+	// Default: 5
+	// +optional
+	KubeAPIQPS *int32 `json:"kubeAPIQPS,omitempty"`
+	// kubeAPIBurst is the burst to allow while talking with kubernetes apiserver
+	// Default: 10
+	// +optional
+	KubeAPIBurst int32 `json:"kubeAPIBurst,omitempty"`
 	// serializeImagePulls when enabled, tells the Kubelet to pull images one
 	// at a time. We recommend *not* changing the default value on nodes that
 	// run docker daemon with version  < 1.9 or an Aufs storage backend.
 	// Issue #10959 has more details.
-	SerializeImagePulls *bool `json:"serializeImagePulls"`
+	// Default: true
+	// +optional
+	SerializeImagePulls *bool `json:"serializeImagePulls,omitempty"`
 	// Map of signal names to quantities that defines hard eviction thresholds. For example: {"memory.available": "300Mi"}.
+	// Default:
```

Comment:
If this is set to omitempty, then this object can't be serialized correctly to stored form.  omitempty ensures that empty arrays are set to nil, which is used as a default.  It looks like today we support empty map on this - how would I specify that in the future?

Reply:
Hmm. I should probably go back through and audit this for all the fields that have collection types, just to be safe.

Reply:
I went through the fields, table below. The only one I'm sure needs to change is, as @smarterclayton noted, `EvictionHard`.

I asked @tallclair about the `Host*Sources` fields, and the opinion is that they provide a false sense of security, and unnecessarily duplicate security config in the Kubelet config, and likely we should remove them from the config type before moving it to beta.

I also asked @tallclair to read through the type struct and note any other fields that he thinks should be removed. 

### Theory
Omitempty ensures that empty container types (slices and maps) are set to nil when they are serialized. 

If an empty container type is a valid value for a field, and is not the default value for that field, we cannot use omitempty on that field; it will be omitted when serialized to disk, and the non-empty default would override when deserialized and run through the defaulter.

If an empty array or map is a valid value, but is also the default for the field, then we can use omitempty, because omitting the empty container will have the same semantic as providing it.
We should try to use omitempty when possible, because it results in more compact serializations. When we don't use omitempty, we should provide a comment that explains why.

### Audit
Field | Type | Default | Empty is valid | omitempty | must change from current
-- | -- | -- | -- | -- | --
ManifestURLHeader | map[string][]string | nil | yes | yes | no
HostNetworkSources | []string | ["*"] | **?** | **?** | **?**
HostPIDSources | []string | ["*"] | **?** | **?** | **?**
HostIPCSources | []string | ["*"] | **?** | **?** | **?**
ClusterDNS | []string | nil | yes | yes | no
EvictionHard | map[string]string | non-nil map | yes | no | **yes**
EvictionSoft | map[string]string | nil | yes | yes | no
EvictionSoftGracePeriod | map[string]string | nil | yes | yes | no
EvictionMinimumReclaim | map[string]string | nil | yes | yes | no
SystemReserved | map[string]string | nil | yes | yes | no
KubeReserved | map[string]string | nil | yes | yes | no
EnforceNodeAllocatable | []string | ["pods"] | yes | no | no

Reply:
treating nil and [] as different (as EnforceNodeAllocatable does) is fraught... this bit us badly when we added proto serializations that did not distinguish between the two